lib/server_sent_event_stage.ex

```diff
@@ -52,79 +53,123 @@ defmodule ServerSentEventStage do
-    state = %{state | state: :connected}
+  def handle_info(message, %{conn: conn} = state) when conn != nil do
+    case HTTP.stream(state.conn, message) do
+      {:ok, conn, responses} ->
+        state = %{state | conn: conn}
+        {state, events} = Enum.reduce_while(responses, {state, []}, &handle_mint_response/2)
 
-    state =
-      if code == 200 do
-        Logger.debug(fn -> "#{__MODULE__} connected" end)
-        state
-      else
-        Logger.warn(fn -> "#{__MODULE__} unexpected status: #{code}" end)
-        do_refresh(state)
-      end
+        {:noreply, events, state}
+
+      {:error, conn, error, responses} ->
+        state = %{state | conn: conn}
+        ref = state.ref
+        {state, events} = Enum.reduce_while(responses, {state, []}, &handle_mint_response/2)
+        {_, {state, events}} = handle_mint_response({:error, ref, error}, {state, events})
+
+        {:noreply, events, state}
+
+      :unknown ->
+        handle_unknown_info(message, state)
+    end
+  end
+
+  def handle_info(message, state) do
+    handle_unknown_info(message, state)
+  end
 
+  def handle_unknown_info({data_tag, _, _}, state) when data_tag in [:ssl, :tcp] do
+    # The can occur after we've re-connected: drop them on the floor
     {:noreply, [], state}
   end
 
-  def handle_info(%HTTPoison.AsyncHeaders{id: id}, %{id: id} = state) do
+  def handle_unknown_info({closed_tag, _}, state) when closed_tag in [:ssl_closed, :tcp_closed] do
+    # These can occur after we've re-connected: drop them on the floor.
     {:noreply, [], state}
   end
 
-  def handle_info(
-        %HTTPoison.AsyncChunk{id: id, chunk: chunk},
-        %{id: id, state: :connected} = state
-      ) do
+  def handle_unknown_info(message, state) do
+    # ignore data received unexpectedly
+    Logger.warn(fn ->
+      "#{__MODULE__} unexpected message: #{inspect(message)}\nState: #{inspect(state)}"
+    end)
+
+    {:noreply, [], state}
+  end
+
+  defp handle_mint_response({:status, ref, 200}, {%{ref: ref}, _events} = acc) do
+    Logger.debug(fn -> "#{__MODULE__} connected" end)
+    {:cont, acc}
+  end
+
+  defp handle_mint_response({:status, ref, redirect_code}, {%{ref: ref} = state, events})
+       when redirect_code in [301, 302, 307] do
+    Logger.debug(fn -> "#{__MODULE__} connected, received redirect #{redirect_code}" end)
+    state = %{state | redirecting?: true}
+    {:cont, {state, events}}
+  end
+
+  defp handle_mint_response({:status, ref, code}, {%{ref: ref} = state, events}) do
+    Logger.warn(fn -> "#{__MODULE__} unexpected status: #{code}" end)
+    state = do_refresh(state)
+    {:halt, {state, events}}
+  end
+
+  defp handle_mint_response(
+         {:headers, ref, _headers},
+         {%{ref: ref, redirecting?: false}, _events} = acc
+       ) do
+    {:cont, acc}
+  end
+
+  defp handle_mint_response(
+         {:headers, ref, headers},
+         {%{ref: ref, redirecting?: true} = state, events}
```

Comment:
Does the `:status` response always come before the `:headers` response? If I understand correctly, we need it to in order to set `:redirecting?` to true for this clause to catch. If it could come in either order we'd maybe need to keep a `:new_location` state or something which the headers can set, and then if we get `:status` of 301 afterwards we do this same reconnect dance.

Reply:
It does: HTTP statuses come before the headers in the protocol.